src/librados/librados_cxx.cc

```diff
@@ -109,27 +112,35 @@ void librados::ObjectOperation::set_op_flags(ObjectOperationFlags flags)
   set_op_flags2((int)flags);
 }
 
+#define MAYBE_RECREATE \
+  if (unlikely(!impl))				\
+    impl = new librados::ObjectOperationImpl
```

Comment:
@adamemerson can we just throw an exception in this case instead of lazily creating a new `librados::ObjectOperationImpl` for it ? is there any case we still expect the `ObjectOperation` instance is still usable after it is moved away?

Reply:
Nothing expects it, since it wasn't movable before. My reasoning is that at present nothing in librados ever throws an exception (or otherwise fatal behavior) so given that I'm adding new behavior it's better to do so in a way that doesn't add one corner where we throw exceptions.

Reply:
@adamemerson i'd prefer `ceph_assert(impl)` then, if this is unprecedented. IMO, it sounds like a bug if we want to use an "empty" object.

Reply:
`ceph_assert()` is a good compromise here, yeah. we might also want to document that these are invalid after move until you reassign a valid instance, and maybe provide an `operator bool()` to test whether it's valid

Reply:
I'm not sure why there's a preference for assert here? We'd be paying the cost of a conditional either way, whether we're lazily allocating a new ObjectOperation if the user tries to use one or testing to abort. And the latter avoids having to introduce a concept of an 'invalid object'.

Reply:
@adamemerson i think that the concept of an "invalid object" is inevitable if we *transfer* the innards from an object to another. and it's more correct than lazily allocating a new ObjectOperation on demand in the sense that 

- it's expected behavior of `std::move()`
- that said, if we don't expect this behavior -- a usable object even it's moved away, it's a bug if we rely on this.

 the reason why i like assert() here is that it helps the developer to identify the bug of using an "invalid object" instead of taking time looking into a segfault issue.

Reply:
in general, making assumptions about the state of a moved-from object seems like a mistake that we should try to mimimize in design. considering some generic type T, you don't even know whether it implements move construction or assignment, so attempts to use them may just fall back to copies. so i'd much rather have a step like reset() or operator=() that makes this reinitialization explicit, rather than relying on an implicit assumption that it's safe for some given type

you're right that the overhead is the same either way; but in the ideal case these asserts would compile out of release builds, and we'd only pay a cost on reset()/operator=() for the cases that need it - while debug builds would still catch the mistakes early on in development

ultimately this is just my preference, and i'm fine either way as long as the move semantics are documented clearly